### PR TITLE
Patches to the build system to properly support importing ioda-converters

### PR DESCRIPTION
**Pull request by rhoneyager-tomorrow**
## Description

This PR makes a few small fixes to the ioda-converters build system to allow it to be found using `find_package(iodaconv)` by dependent packages. I've fixed a few typos, and `make install` now works correctly. 

## Issue(s) addressed

None noted in JCSDA-internal, but this fixes some bugs on my end. Now I can reference `bufr2ioda.x` and related apps in my CMake projects.

I've made `ioda` a required package. CMake already threw cryptic errors when it wasn't found, and it doesn't really make sense to build ioda-converters without ioda anymore. Too many programs, scripts, and libraries depend on it.

## Dependencies

None

## Impact

Mildly positive.

## Checklist

- [x] I have performed a self-review of my own code
    - The changes work on my end.
- [x] I have made corresponding changes to the documentation
- [x] I have run the unit tests before creating the PR
    - The unit tests cannot test this capability, but I've tested manually with a project that runs `find_package(iodaconv REQUIRED)` and references `bufr2ioda.x`. It works.

### CODE REVIEW

File: CMakeLists.txt
Comment (rhoneyager-tomorrow) [inline CMakeLists.txt]: ecbuild takes this as a hint when creating new targets.

File: CMakeLists.txt
Comment (rhoneyager-tomorrow) [inline CMakeLists.txt:21]: iodaconv was using ecbuild instructions without it being declared as an ecbuild project.

File: CMakeLists.txt
Comment (rhoneyager-tomorrow) [inline CMakeLists.txt]:
```suggestion
```

**Comment by BenjaminRuston** [comment]
@eap just putting you on this for awareness anyone else (@fmahebert ?) to bring in just to make sure of consistency between the repos

File: CMakeLists.txt
Comment (srherbener) [reply CMakeLists.txt]: @rhoneyager-tomorrow are you saying that ecbuild uses `INSTALL_LIB_DIR` as a hint? If so, wouldn't this need to be aware of which platforms use `lib` and which use `lib64`? If this is a valid concern, I think it would be okay to defer solving this to a later PR (ie, coordinated with other JEDI repos).

File: CMakeLists.txt
Comment (srherbener) [reply CMakeLists.txt:21]: Good catch! Thanks

**Review by srherbener — COMMENTED** [review]
Thanks for these fixes! They are much appreciated.

**Comment by rhoneyager-tomorrow** [comment]
> @rhoneyager-tomorrow are you saying that ecbuild uses INSTALL_LIB_DIR as a hint?

Yes, ecbuild uses `INSTALL_BIN_DIR` and `INSTALL_LIB_DIR` as hints to set these paths. See [ecbuild_add_library](https://github.com/ecmwf/ecbuild/blob/develop/cmake/ecbuild_add_library.cmake#L499) for examples for how these are translated into the lower-level CMake calls.

> If so, wouldn't this need to be aware of which platforms use lib and which use lib64? If this is a valid concern, I think it would be okay to defer solving this to a later PR (ie, coordinated with other JEDI repos).

You can set the ecbuild hint if you want, or you can scratch out that line. It's cosmetic.

Each Linux / Unix distribution adopts their own standard, and CMake makes no attempt to match a specific system. The CMake defaults are instead taken from the [GNU Coding Standards](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html). For completeness, it's worth mentioning that Fedora and Debian derivatives follow their own variants of the [Filesystem Hierarchy Standard](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) instead.

**Comment by srherbener** [comment]
> > @rhoneyager-tomorrow are you saying that ecbuild uses INSTALL_LIB_DIR as a hint?
> 
> Yes, ecbuild uses `INSTALL_BIN_DIR` and `INSTALL_LIB_DIR` as hints to set these paths. See [ecbuild_add_library](https://github.com/ecmwf/ecbuild/blob/develop/cmake/ecbuild_add_library.cmake#L499) for examples for how these are translated into the lower-level CMake calls.
> 
> > If so, wouldn't this need to be aware of which platforms use lib and which use lib64? If this is a valid concern, I think it would be okay to defer solving this to a later PR (ie, coordinated with other JEDI repos).
> 
> You can set the ecbuild hint if you want, or you can scratch out that line. It's cosmetic.
> 
> Each Linux / Unix distribution adopts their own standard, and CMake makes no attempt to match a specific system. The CMake defaults are instead taken from the [GNU Coding Standards](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html). For completeness, it's worth mentioning that Fedora and Debian derivatives follow their own variants of the [Filesystem Hierarchy Standard](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) instead.

@rhoneyager-tomorrow thanks for the response. It looks like the `ecbuild_declare_project` macro sets `INSTALL_LIB_DIR` to "lib" as the default value. Ie, the setting of `INSTALL_LIB_DIR` to "lib" in this script is redundant, and if the user needs to change this setting it can be done on the ecbuild command line instead of having to edit this script. Also, if the set command is left here and the user changes INSTALL_LIB_DIR to "lib64" on the ecbuild command line (and forgets to update the setting in this script) then INSTALL_LIB_DIR will get reset to "lib" for this repo and subsequent repos.

 It seems to me, unless I'm missing something, that it would be cleaner to remove the `set ( INSTALL_LIB_DIR lib )` statement and handle variants such as "lib64" on the command line (eg. `ecbuild -DINSTALL_LIB_DIR=lib64 ...`). That way, the INSTALL_LIB_DIR setting will be consistent across all the repos. Would that be an acceptable practice?

Thanks!

**Comment by rhoneyager-tomorrow** [comment]
Yeah, I agree. I've removed the line.

**Review by srherbener — APPROVED** [review]
Thanks for addressing my comments! You help with this is very much appreciated.

**Review by fmahebert — APPROVED** [review]
Mine is a non-expert review, but these changes look sensible to me.

Thanks @rhoneyager-tomorrow !

**Comment by rhoneyager-tomorrow** [comment]
A few weeks have passed since approval. Is anything blocking this PR from being merged?

**Comment by srherbener** [comment]
> A few weeks have passed since approval. Is anything blocking this PR from being merged?

Sorry @rhoneyager-tomorrow for the delay. I'm not the repo "owner", but I've got the permissions to do the merge. I've just synced this PR with develop and I'll do the merge once the tests run (and pass).

**Comment by srherbener** [comment]
@eap, the CI unit test has been spinning its wheels since yesterday afternoon. Any ideas what is wrong? Also, I'm not sure why the integration test is skipped. Thanks!

**Comment by eap** [comment]
@srherbener - If found the following error in the test log. there's a checkout issue with a bad commit reference (is it possible that a forced push was used overwriting a branch, or maybe this is a transient network error).

Also a second issue that might be related or unrelated.

```
-- Use test data from ioda-data repository
-- ---------------------------------------------------------
-- Adding bundle project iodaconv
-- Cloning iodaconv from https://github.com/jcsda-internal/ioda-converters.git into /workdir/test_root/jedi-bundle/iodaconv...
...
  CRITICAL - git checkout 468d688d9db5b279fa63753abd9ffac353ee8d95 on
  /workdir/test_root/jedi-bundle/iodaconv failed:
    /usr/bin/git checkout -q 468d688d9db5b279fa63753abd9ffac353ee8d95
    fatal: reference is not a tree: 468d688d9db5b279fa63753abd9ffac353ee8d95
...

CMake Error: CTEST_USE_LAUNCHERS is enabled, but the RULE_LAUNCH_COMPILE global property is not defined.
Did you forget to include(CTest) in the toplevel CMakeLists.txt ?-- Configuring incomplete, errors occurred!
See also "/workdir/test_root/build/CMakeFiles/CMakeOutput.log".
See also "/workdir/test_root/build/CMakeFiles/CMakeError.log".

```

**Comment by rhoneyager-tomorrow** [comment]
Looks like the CI system is choking since my PR comes from my fork, and that's because I no longer have write access on JCSDA-internal. This is a bug in how ecbuild and the CI system are interacting. See https://github.com/ecmwf/ecbuild/blob/develop/cmake/ecbuild_git.cmake#L99. CI is first cloning the JCSDA-internal/ioda-converters default branch and then trying to checkout the commit, but that commit is on a different remote.

The quick fix would be for you to make a local copy of my branch and resubmit this PR.

**Review by RussTreadon-NOAA — APPROVED** [review]
Approve.  

The changes in this PR are required for GDASApp branch [`feature/install`](https://github.com/NOAA-EMC/GDASApp/tree/feature/install) to successfully complete the install (see [comment here](https://github.com/NOAA-EMC/GDASApp/issues/1302#issuecomment-2394398595)).  

Without the changes in this branch, `feature/install` winds up with
```
(gdasapp) hercules-login-2:/work/noaa/da/rtreadon/git/GDASApp/install_lib/install$ ls
'$(PYIODA_INSTALL_LIBDIR)'        bin   include   lib64    MOM6    test
'$(PYIOODACONV_INSTALL_LIBDIR)'   doc   lib       module   share   ush
```
(See GDASApp issue[ #1302](https://github.com/NOAA-EMC/GDASApp/issues/1302) [comment here](https://github.com/NOAA-EMC/GDASApp/issues/1302#issuecomment-2394309028))

**Comment by BenjaminRuston** [comment]
@fmahebert , @RussTreadon-NOAA, @rhoneyager-tomorrow  and @srherbener 

this PR has moved to:  https://github.com/JCSDA-internal/ioda-converters/pull/1566

and will close this as a duplicate once that is merged

**Comment by BenjaminRuston** [comment]
closed as duplicate

**Pull request closed without merging**
